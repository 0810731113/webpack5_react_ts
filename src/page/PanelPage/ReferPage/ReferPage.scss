@import "style/colors.scss";
.refer-title {
  font-size: 12px;
  color: $text-color-secondary;
}
.ant-row + .ant-row {
  margin-top: 10px;
}
.action-row {
  justify-content: space-between;
}
.filter-row {
  .type-filter {
    flex: auto;
    margin-left: 4px;
  }
}
.refer-dropdown {
  &.ant-dropdown-placement-bottomRight {
    .ant-dropdown-arrow {
      border-left-color: #001529;
      border-top-color: #001529;
    }
  }
}
