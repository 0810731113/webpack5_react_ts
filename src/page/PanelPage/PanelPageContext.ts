import { ResourcePermission } from "api-authorization/generated/model";
import { AxiosError } from "axios";
import { createContext } from "react";
import {
  DataSetVO,
  SpecialtyVO,
  SpecialtyVOTypeEnum,
  VersionVO,
} from "api/generated/model";
import { ReferedVersionVO } from "./ReferPage";

export enum ProjectContentEnum {
  Samples = "samples",
  Mine = "",
}

export interface PanelPageContext {
  userId: string | null;
  accessToken: string | null;
  workUnit: DataSetVO | null;
  workUnits: DataSetVO[] | null;
  version: VersionVO | null;
  projectId: string | null;
  specialtyType: SpecialtyVOTypeEnum;
  specialties?: SpecialtyVO[] | null;
  referingVersionIds: number[];
  referedVersions: ReferedVersionVO[];
  referWorkUnits: DataSetVO[] | null;
  workunitReadonly: boolean;
  showprocess: 0 | 1;
  logs: any[];
  resources: ResourcePermission[];
  projectContent: ProjectContentEnum | null;
  setSpecialties(value: SpecialtyVO[] | null | undefined): void;

  currentOperatingWorkUnitId: string | null;
  // 打开工作单元
  openingWorkUnit: boolean;
  setOpeningWorkUnit(value: boolean): void;

  // 保存工作单元
  savingWorkUnit: boolean;
  setSavingWorkUnit(value: boolean): void;

  // 提交工作单元
  commitingWorkUnit: boolean;
  setCommitingWorkUnit(value: boolean): void;
  setShowprocess(value: 0 | 1): void;

  setWorkUnit: (workUnit: DataSetVO | null) => void;
  setWorkUnits: (workUnits: DataSetVO[] | null) => void;
  setReferWorkUnits: (workUnits: DataSetVO[] | null) => void;
  setVersion: (version: VersionVO | null) => void;
  setCurrentOperationWorkUnitId: (id: string | null) => void;
  setProjectId: (id: string | null) => void;
  setReferingVersionIds: (ids: number[]) => void;
  setReferedVersions: (versions: ReferedVersionVO[]) => void;
  setWorkunitReadonly: (readonly: boolean) => void;
  pushLog: (log: any) => void;
  resetProject: () => void;
  setProjectContent: (content: ProjectContentEnum) => void;

  onResponseError: (error: AxiosError) => void;

  refreshCount: number;
}

export default createContext<PanelPageContext>({} as any);
