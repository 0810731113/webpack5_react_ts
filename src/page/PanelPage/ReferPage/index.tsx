/* eslint-disable @typescript-eslint/no-unused-expressions */
/* eslint-disable default-case */
/* eslint-disable @typescript-eslint/no-shadow */
import React, { useContext, useEffect, ReactText } from "react";
import PanelTabTitle from "component/PanelTabTitle";
import { teamService } from "service";
import { remove } from "lodash";
import { useImmer } from "use-immer";
import {
  VersionVO,
  DataSetVO,
  Team,
  ShareRecordVOStatusEnum,
} from "api/generated/model";
import FakeProgress from "component/FakeProgress";
import panelService from "service/panel.service";
import WorkUnitTree from "component/WorkUnitTree";
import Loading from "component/Loading";
import {
  CompassFilled,
  EyeOutlined,
  EyeInvisibleOutlined,
} from "@ant-design/icons";
import {
  useWorkUnitListByProjectId,
  useVersionListByWorkUnits,
} from "hook/use-work-unit-service.hook";
import { usePermissions } from "hook/use-permission.hook";
import { useShareRecordsWithoutUserName } from "hook/use-share-service.hook";
import { version } from "punycode";
import { Dropdown, Menu, message, Button } from "antd";
import qwebService from "service/qweb.service";
import { publishEvent } from "function/stats.func";
import { specialtyTypeName } from "AppPanel";
import { QTResponse } from "../model";
import ReferDetail from "./ReferDetail";
import PanelPageContext from "../PanelPageContext";

export interface ReferedVersionVO extends VersionVO {
  checkedDataCodeKeys?: ReactText[];
}
interface ReferPageProps {
  activeKey: string;
}

interface ReferedData {
  datasetid: string;
  version: number;
  linkedtypename: string;
  datacodes: string[];
  visible: boolean;
}
interface referedDataResponse {
  linkedworkunits: ReferedData[];
}

interface ReferPageState {
  teams: Team[];
  versions: { [workUnitId: string]: VersionVO[] } | null;
  selectedVersions: { [workUnitId: string]: VersionVO | null };
  datasets: { [teamId: string]: DataSetVO[] } | null;
  openedTeamIds: string[];
  currentDataSet?: DataSetVO;
  currentVersion?: ReferedVersionVO;
  referedVersions: ReferedVersionVO[];
  hidedDataSetIds: string[];
}
export default function ReferPage(props: ReferPageProps) {
  const { activeKey } = props;
  const [
    {
      versions,
      datasets,
      teams,
      currentDataSet,
      currentVersion,
      referedVersions,
      hidedDataSetIds,
    },
    updateState,
  ] = useImmer<ReferPageState>({
    versions: null,
    datasets: null,
    selectedVersions: {},
    teams: [],
    openedTeamIds: [],
    referedVersions: [],
    hidedDataSetIds: [],
  });
  const {
    projectId,
    workUnit,
    referingVersionIds,
    refreshCount,
    workunitReadonly,
    specialtyType,
    setReferedVersions,
    setReferingVersionIds,
    setReferWorkUnits,
  } = useContext(PanelPageContext);
  const {
    workUnits: allWorkUnits,
    loading,
    refresh: refreshAllWorkUnit,
  } = useWorkUnitListByProjectId(projectId ?? undefined);
  const {
    records: consumedData,
    loading: loadingConsumedData,
    refresh: refreshShareRecords,
  } = useShareRecordsWithoutUserName({
    consumeId: workUnit?.teamId!,
    status: [ShareRecordVOStatusEnum.Consumed],
  });
  const {
    permissions,
    loading: loadingPermissions,
    refresh: refreshPermissions,
  } = usePermissions(workUnit?.teamId!, "TrustBy");
  const {
    versions: allVersions,
    loading: loadingVersions,
    refresh: refreshVersions,
  } = useVersionListByWorkUnits(allWorkUnits, "committed");

  useEffect(() => {
    updateState((draft) => {
      draft.referedVersions = [];
      draft.hidedDataSetIds = [];
    });
  }, [workUnit?.id]);
  useEffect(() => {
    const subscription = qwebService.subscribe((e) => {
      console.log("received qt event: ", e);
      switch (e.CollaborationClientEvent.actionname) {
        case "linkedworkunitdeleted":
          // message.info("工作单元已关闭");
          updateState((draft) => {
            draft.referedVersions = referedVersions.filter(
              (referedVersion) =>
                referedVersion.dataSetId !==
                e.CollaborationClientEvent.arguments?.datasetid,
            );
          });
          break;
        default:
          break;
      }
    });
    setReferedVersions(referedVersions);
    return () => {
      subscription?.unsubscribe();
    };
  }, [referedVersions]);
  const prepareData = async () => {
    if (projectId) {
      const newTeams = await teamService.listTeams(projectId);
      updateState((draft) => {
        draft.teams = newTeams!;
        draft.openedTeamIds = newTeams ? newTeams.map((team) => team.id!) : [];
      });
    }
  };

  useEffect(() => {
    if (workUnit?.id) {
      prepareData();
    }
  }, [projectId, workUnit?.id]);
  const init = () => {
    prepareData();
    refreshVersions();
    refreshAllWorkUnit();
    refreshPermissions();
    refreshShareRecords();
  };
  const refreshReferedVersions = () => {
    workUnit &&
      allVersions &&
      panelService.getlinkedworkunits(workUnit?.id!).then((data: any) => {
        updateState((draft) => {
          draft.referedVersions = [
            ...referedVersions,
            ...(data?.linkedworkunits
              ?.filter(
                (item: ReferedData) =>
                  !referedVersions.some(
                    (version) => version?.dataSetId === item.datasetid,
                  ),
              )
              .map((item: ReferedData) => ({
                ...(allVersions.find(
                  (version) =>
                    version.dataSetId === item.datasetid &&
                    version.version === item.version,
                ) || {}),
                dataSetId: item.datasetid,
                checkedDataCodeKeys: item.datacodes,
              })) ?? []),
          ];
          draft.hidedDataSetIds = [
            ...hidedDataSetIds,
            ...(data?.linkedworkunits
              ?.filter(
                (item: ReferedData) =>
                  !hidedDataSetIds.some(
                    (dataSetId) => dataSetId === item.datasetid,
                  ) && !item.visible,
              )
              .map((item: ReferedData) => item.datasetid) ?? []),
          ];
        });
      });
  };
  // useEffect(() => {
  //   if (activeKey === "2") {
  //     try {
  //       init();
  //     } catch (err) {
  //       message.error(err);
  //     }
  //   }
  // }, [activeKey]);
  useEffect(() => {
    try {
      init();
    } catch (err) {
      message.error(err);
    }
  }, [refreshCount, workUnit?.id, activeKey]);
  useEffect(() => {
    try {
      refreshReferedVersions();
    } catch (err) {
      message.error(err);
    }
  }, [workUnit?.id, allVersions]);
  useEffect(() => {
    if (
      teams?.length > 0 &&
      consumedData &&
      allWorkUnits?.length > 0 &&
      permissions &&
      allVersions?.length > 0
    ) {
      updateState((draft) => {
        const selectWorkUnits = allWorkUnits.filter(
          (unit) =>
            !(
              unit.name === workUnit?.name && unit.teamId === workUnit?.teamId
            ) && // Todo 排除自己
            !(
              unit.type === "workunit" ||
              (unit.teamId !== workUnit?.teamId &&
                !permissions.some(
                  (permission) => permission.teamId === unit.teamId,
                ) &&
                !consumedData.some((data) =>
                  data.contents?.some(
                    (content) => content.dataSetId === unit.id,
                  ),
                ))
            ),
        );
        setReferWorkUnits(selectWorkUnits);
        draft.datasets = selectWorkUnits.reduce(
          (result: { [teamId: string]: DataSetVO[] }, workUnit) => {
            if (!result[workUnit.teamId!]) {
              result[workUnit.teamId!] = [];
            }
            result[workUnit.teamId!].push(workUnit);
            return result;
          },
          {},
        );
        const initVersions: { [workUnitId: string]: VersionVO[] } = {};
        draft.versions = selectWorkUnits.reduce(
          (result, unit) => ({
            ...result,
            [unit.id!]: allVersions.filter(
              (version) =>
                version?.version &&
                version?.dataSetId === unit.id &&
                (unit.teamId !== workUnit?.teamId &&
                !permissions.some(
                  (permission) => permission.teamId === unit.teamId,
                )
                  ? consumedData?.some((data) =>
                      data.contents?.some(
                        (content) => version.id === content.id,
                      ),
                    )
                  : true),
            ),
          }),
          initVersions,
        );
      });
    }
  }, [teams, consumedData, allWorkUnits, permissions, allVersions]);
  const onReferCommit = (checkedKeys: ReactText[]) => {
    const findedIndex = referedVersions.findIndex(
      (version) =>
        version.id === currentVersion?.id ||
        (version?.dataSetId === currentVersion?.dataSetId &&
          version?.version === currentVersion?.version),
    );

    if (checkedKeys && checkedKeys.length) {
      publishEvent("referWorkUnit", ["工具"], {
        from: specialtyTypeName[specialtyType],
        query: checkedKeys.includes("all") ? "全部" : "部分",
      });
      setReferingVersionIds([...referingVersionIds, currentVersion?.id!]);
      panelService
        .linkworkunit(
          currentDataSet?.id!,
          currentVersion?.version!,
          currentDataSet?.name!,
          Array.from(new Set(checkedKeys.join(",").split(","))),
        )
        .then((result) => {
          const newReferedVersion = {
            ...currentVersion,
            checkedDataCodeKeys: checkedKeys,
          };
          updateState((draft) => {
            if (findedIndex === -1) {
              draft.referedVersions = [...referedVersions, newReferedVersion];
            } else {
              draft.referedVersions = referedVersions.map((referedVersion, i) =>
                i === findedIndex ? newReferedVersion : referedVersion,
              );
            }
            draft.hidedDataSetIds = hidedDataSetIds.filter(
              (dataSetId) => dataSetId !== currentVersion?.dataSetId!,
            );
          });
          setReferingVersionIds(
            referingVersionIds.filter(
              (versionId) => versionId !== currentVersion?.id!,
            ),
          );
        });
    } else if (findedIndex !== -1) {
      updateState((draft) => {
        draft.referedVersions = referedVersions.filter(
          (version, index) => index !== findedIndex,
        );
      });
    }
    updateState((draft) => {
      draft.currentDataSet = undefined!;
      draft.currentVersion = undefined;
    });
  };
  const onActionClick = (
    key: string,
    dataSet: DataSetVO,
    version: ReferedVersionVO,
  ) => {
    switch (key) {
      case "update":
        if (versions) {
          panelService
            .updatelinkedworkunit(
              dataSet.id!,
              versions[dataSet.id!][0]?.version!,
            )
            .then(() => {
              updateState((draft) => {
                const newReferedVersions = referedVersions.filter(
                  (referedVersion) =>
                    referedVersion.id !== version.id &&
                    !(
                      version?.dataSetId === referedVersion?.dataSetId &&
                      version?.version === referedVersion?.version
                    ),
                );
                newReferedVersions.push(versions[dataSet.id!][0]);
                draft.referedVersions = newReferedVersions;
                draft.versions = {
                  ...versions,
                  [dataSet.id!]: [...versions[dataSet.id!]],
                };
              });
            });
        }
        break;
      case "reSelect":
        updateState((draft) => {
          draft.currentDataSet = dataSet!;
          draft.currentVersion = {
            ...version,
            checkedDataCodeKeys: referedVersions.find(
              (v) => v.dataSetId === dataSet.id,
            )?.checkedDataCodeKeys,
          };
        });
        break;
      case "delete":
        panelService.deletelinkedworkunit(dataSet.id!).then(() => {
          updateState((draft) => {
            draft.referedVersions = referedVersions.filter(
              (referedVersion) =>
                referedVersion.id !== version.id &&
                !(
                  version?.dataSetId === referedVersion?.dataSetId &&
                  version?.version === referedVersion?.version
                ),
            );
          });
        });
        break;
    }
  };
  if (loading || loadingConsumedData || loadingPermissions || loadingVersions) {
    return <Loading absolute />;
  }
  return (
    <div className="panel-body">
      <div className="title">
        <PanelTabTitle
          title="可选参照"
          tip="你可以通过链接一个其他人的工作单元，作为自己的设计参照"
        />
      </div>
      <WorkUnitTree
        teams={teams}
        datasets={datasets}
        versions={versions}
        referedVersions={referedVersions}
        versionDisabled={(dataSet, version) =>
          referedVersions.some(
            (referedVersion) =>
              referedVersion?.id === version?.id ||
              (version?.dataSetId === referedVersion?.dataSetId &&
                version?.version === referedVersion?.version),
          )
        }
        nodeSelected={(dataSet, version) =>
          referedVersions.some(
            (referedVersion) =>
              referedVersion?.id === version?.id ||
              (version?.dataSetId === referedVersion?.dataSetId &&
                version?.version === referedVersion?.version),
          )
        }
        preIconRender={(dataSet, version) => {
          const findedReferedVersion = referedVersions.find(
            (referedVersion) =>
              referedVersion.id === version?.id ||
              (version?.dataSetId === referedVersion?.dataSetId &&
                version?.version === referedVersion?.version),
          );
          return (
            !referingVersionIds.includes(version?.id!) &&
            findedReferedVersion &&
            (hidedDataSetIds.includes(dataSet?.id!) ? (
              <EyeInvisibleOutlined
                onClick={async () => {
                  await panelService.showlinkedworkunit(dataSet.id!);
                  updateState((draft) => {
                    draft.hidedDataSetIds = hidedDataSetIds.filter(
                      (id) => dataSet.id !== id,
                    );
                  });
                }}
              />
            ) : (
              <EyeOutlined
                onClick={async () => {
                  await panelService.hidelinkedworkunit(dataSet.id!);
                  updateState((draft) => {
                    draft.hidedDataSetIds = [...hidedDataSetIds, dataSet.id!];
                  });
                }}
              />
            ))
          );
        }}
        bottomActionRender={(dataSet, version) => {
          const findedReferedVersion = referedVersions.find(
            (referedVersion) =>
              referedVersion.id === version?.id ||
              (version?.dataSetId === referedVersion?.dataSetId &&
                version?.version === referedVersion?.version),
          );
          const maxVersion =
            version?.dataSetId && versions?.[version?.dataSetId]?.[0]?.version;
          const nowVersion = version?.version;
          return (
            !referingVersionIds.includes(version?.id!) &&
            findedReferedVersion && (
              <>
                {maxVersion && nowVersion && maxVersion > nowVersion && (
                  <Button
                    type="link"
                    shape="circle"
                    disabled={workunitReadonly}
                    onClick={() => onActionClick("update", dataSet, version)}
                    size="small"
                  >
                    更新
                  </Button>
                )}
                <Button
                  type="link"
                  disabled={workunitReadonly}
                  shape="circle"
                  onClick={() => onActionClick("reSelect", dataSet, version)}
                  size="small"
                >
                  重新筛选
                </Button>
                <Button
                  type="link"
                  shape="circle"
                  disabled={workunitReadonly}
                  onClick={() => onActionClick("delete", dataSet, version)}
                  size="small"
                >
                  删除
                </Button>
              </>
            )
          );
        }}
        actionRender={(dataSet, version) => {
          const findedReferedVersion = referedVersions.some(
            (referedVersion) =>
              referedVersion.id === version?.id ||
              (version?.dataSetId === referedVersion?.dataSetId &&
                version?.version === referedVersion?.version),
          );
          return referingVersionIds.includes(version?.id!) ? (
            <FakeProgress
              completed={false}
              showInfo={false}
              width={20}
              onCompleted={() => {
                setReferingVersionIds(
                  referingVersionIds.filter(
                    (versionId) => versionId !== version.id,
                  ),
                );
              }}
            />
          ) : (
            <Button
              size="small"
              style={{ fontSize: 12, padding: "0px 4px", minWidth: "auto" }}
              disabled={workunitReadonly || findedReferedVersion}
              onClick={() => {
                updateState((draft) => {
                  draft.currentDataSet = dataSet!;
                  draft.currentVersion = version;
                });
              }}
            >
              参照
            </Button>
          );
        }}
      />
      <ReferDetail
        onCommit={onReferCommit}
        visible={!!currentDataSet && !!currentVersion}
        dataSet={currentDataSet}
        version={currentVersion}
        onClose={() => {
          updateState((draft) => {
            draft.currentDataSet = undefined!;
            draft.currentVersion = undefined;
          });
        }}
      />
    </div>
  );
}
