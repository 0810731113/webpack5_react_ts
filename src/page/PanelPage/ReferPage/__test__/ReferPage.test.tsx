import React from "react";
import pretty from "pretty";
import { render, screen } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { render as renderHtml, unmountComponentAtNode } from "react-dom";

import { teamService } from "service";
import PanelPageContext from "page/PanelPage/PanelPageContext";
import ReferPage from "../index";

let container: HTMLDivElement | null;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  if (container) {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  }
});

jest.mock("service");
jest.mock("consts", () => ({
  isPanel: true,
  PUBLIC_URL: "/web",
}));

test("test refer page", async () => {
  const defaultContext = {
    projectId: "3ff55baf-f4ad-4652-b642-cce9f77f9e2c",
    workUnit: {
      codeVersion: 1,
      description: "",
      folderId: "b4b53e1e-990d-4030-83de-5c4eddce9845",
      id: "b7d12fb2-a04f-41c4-9b91-df646de651b4",
      name: "B",
      nameSpace: null,
      ownerId: "6742694569103397638",
      projectId: null,
      schemaDefinition: null,
      specialtyId: "44d03d60-29bf-45f0-b1f2-b54698d37d0c",
      teamId: "3917d4af-2c29-491c-8c42-9a3961b2d94e",
      tipVersion: 4,
      type: "workunit",
    },
    referingVersionIds: ["38315"],
    refreshCount: 1,
    workunitReadonly: false,
    userId: "6742693964083531997",
    accessToken: "cn-e5a6e077-8602-4f18-ad55-276e099f88e3",
    setReferedVersions() {},
    setReferingVersionIds() {},
    setReferWorkUnits() {},
  };
  await act(async () =>
    renderHtml(
      <div>
        <PanelPageContext.Provider value={defaultContext as any}>
          <ReferPage activeKey="2" />
        </PanelPageContext.Provider>
      </div>,
      container,
    ),
  );

  expect(pretty(container?.innerHTML || "")).toMatchSnapshot();
});
