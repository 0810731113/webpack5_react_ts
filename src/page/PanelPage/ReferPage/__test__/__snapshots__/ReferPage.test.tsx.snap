// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test refer page 1`] = `
"<div>
  <div class=\\"panel-body\\">
    <div class=\\"title\\"><span class=\\"panel-tab-title\\"><span>可选参照：</span><span class=\\"tab-icon-wrap\\"><span role=\\"img\\" aria-label=\\"info-circle\\" class=\\"anticon anticon-info-circle\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"info-circle\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z\\"></path><path d=\\"M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z\\"></path></svg></span></span></span></div>
    <div class=\\"work-unit-tree  \\">
      <div class=\\"team-block \\">
        <div class=\\"team-name\\"><span role=\\"img\\" aria-label=\\"caret-down\\" tabindex=\\"-1\\" class=\\"anticon anticon-caret-down\\"><svg viewBox=\\"0 0 1024 1024\\" focusable=\\"false\\" data-icon=\\"caret-down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z\\"></path></svg></span><span>GAP</span></div>
        <div class=\\"team-content\\">
          <div class=\\"node \\"><span class=\\"pre-icon\\"></span><span class=\\"ant-typography ant-typography-ellipsis\\" style=\\"width: 120px; margin: 0px 0px -4px 0px; overflow: scroll;\\">GMEP</span>
            <div class=\\"node-version\\">
              <div class=\\"ant-select ant-select-sm ant-select-single ant-select-show-arrow\\" style=\\"width: 70px;\\">
                <div class=\\"ant-select-selector\\"><span class=\\"ant-select-selection-search\\"><input autocomplete=\\"off\\" type=\\"search\\" class=\\"ant-select-selection-search-input\\" style=\\"opacity: 0;\\" role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-owns=\\"rc_select_TEST_OR_SSR_list\\" aria-autocomplete=\\"list\\" aria-controls=\\"rc_select_TEST_OR_SSR_list\\" aria-activedescendant=\\"rc_select_TEST_OR_SSR_list_0\\" readonly=\\"\\" unselectable=\\"on\\" value=\\"\\" id=\\"rc_select_TEST_OR_SSR\\"></span><span class=\\"ant-select-selection-item\\">V1</span></div><span class=\\"ant-select-arrow\\" style=\\"user-select: none;\\" unselectable=\\"on\\" aria-hidden=\\"true\\"><span role=\\"img\\" aria-label=\\"down\\" class=\\"anticon anticon-down ant-select-suffix\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z\\"></path></svg></span></span>
              </div>
            </div>
            <div class=\\"node-action\\"><button style=\\"font-size: 12px; padding: 0px 4px; min-width: auto;\\" type=\\"button\\" class=\\"ant-btn ant-btn-sm\\"><span>参 照</span></button></div>
          </div>
          <div class=\\"node \\"><span class=\\"pre-icon\\"></span><span class=\\"ant-typography ant-typography-ellipsis\\" style=\\"width: 120px; margin: 0px 0px -4px 0px; overflow: scroll;\\">A</span>
            <div class=\\"node-version\\">
              <div class=\\"ant-select ant-select-sm ant-select-single ant-select-show-arrow\\" style=\\"width: 70px;\\">
                <div class=\\"ant-select-selector\\"><span class=\\"ant-select-selection-search\\"><input autocomplete=\\"off\\" type=\\"search\\" class=\\"ant-select-selection-search-input\\" style=\\"opacity: 0;\\" role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-owns=\\"rc_select_TEST_OR_SSR_list\\" aria-autocomplete=\\"list\\" aria-controls=\\"rc_select_TEST_OR_SSR_list\\" aria-activedescendant=\\"rc_select_TEST_OR_SSR_list_0\\" readonly=\\"\\" unselectable=\\"on\\" value=\\"\\" id=\\"rc_select_TEST_OR_SSR\\"></span><span class=\\"ant-select-selection-item\\">V1</span></div><span class=\\"ant-select-arrow\\" style=\\"user-select: none;\\" unselectable=\\"on\\" aria-hidden=\\"true\\"><span role=\\"img\\" aria-label=\\"down\\" class=\\"anticon anticon-down ant-select-suffix\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z\\"></path></svg></span></span>
              </div>
            </div>
            <div class=\\"node-action\\"><button style=\\"font-size: 12px; padding: 0px 4px; min-width: auto;\\" type=\\"button\\" class=\\"ant-btn ant-btn-sm\\"><span>参 照</span></button></div>
          </div>
        </div>
      </div>
    </div>
    <div class=\\"\\">
      <div tabindex=\\"-1\\" class=\\"ant-drawer ant-drawer-right issue-detail-drawer refer-detail\\" style=\\"position: absolute;\\">
        <div class=\\"ant-drawer-mask\\"></div>
        <div class=\\"ant-drawer-content-wrapper\\" style=\\"transform: translateX(100%); width: 100%;\\">
          <div class=\\"ant-drawer-content\\">
            <div class=\\"ant-drawer-wrapper-body\\">
              <div class=\\"ant-drawer-header-no-title\\"><button type=\\"button\\" aria-label=\\"Close\\" class=\\"ant-drawer-close\\"><a><span role=\\"img\\" aria-label=\\"arrow-left\\" class=\\"anticon anticon-arrow-left\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"arrow-left\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M872 474H286.9l350.2-304c5.6-4.9 2.2-14-5.2-14h-88.5c-3.9 0-7.6 1.4-10.5 3.9L155 487.8a31.96 31.96 0 000 48.3L535.1 866c1.5 1.3 3.3 2 5.2 2h91.5c7.4 0 10.8-9.2 5.2-14L286.9 550H872c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z\\"></path></svg></span>返回</a></button></div>
              <div class=\\"ant-drawer-body\\">
                <div class=\\"ant-row refer-title\\"><span>正在链接：</span><span></span><span>.</span><span>.V</span></div>
                <div class=\\"ant-row action-row\\"><span class=\\"panel-tab-title\\"><span>选择参照构件：</span><button disabled=\\"\\" type=\\"button\\" class=\\"ant-btn ant-btn-primary ant-btn-sm\\"><span>完 成</span></button></span></div>
                <div class=\\"ant-row filter-row\\"><span class=\\"filter-label\\">筛选：</span>
                  <div class=\\"ant-select ant-select-sm type-filter ant-select-single ant-select-show-arrow\\">
                    <div class=\\"ant-select-selector\\"><span class=\\"ant-select-selection-search\\"><input autocomplete=\\"off\\" type=\\"search\\" class=\\"ant-select-selection-search-input\\" style=\\"opacity: 0;\\" role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-owns=\\"rc_select_TEST_OR_SSR_list\\" aria-autocomplete=\\"list\\" aria-controls=\\"rc_select_TEST_OR_SSR_list\\" aria-activedescendant=\\"rc_select_TEST_OR_SSR_list_0\\" readonly=\\"\\" unselectable=\\"on\\" value=\\"\\" id=\\"rc_select_TEST_OR_SSR\\"></span><span class=\\"ant-select-selection-item\\" title=\\"按构件类型筛选\\">按构件类型筛选</span></div><span class=\\"ant-select-arrow\\" style=\\"user-select: none;\\" unselectable=\\"on\\" aria-hidden=\\"true\\"><span role=\\"img\\" aria-label=\\"down\\" class=\\"anticon anticon-down ant-select-suffix\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z\\"></path></svg></span></span>
                  </div>
                </div>
                <div class=\\"ant-row\\">123<div class=\\"ant-tree ant-tree-icon-hide ant-tree-unselectable\\">
                    <div role=\\"tree\\"><input style=\\"width: 0px; height: 0px; display: flex; overflow: hidden; opacity: 0; border: 0px; padding: 0px; margin: 0px;\\" tabindex=\\"0\\" value=\\"\\"></div>
                    <div class=\\"ant-tree-treenode\\" aria-hidden=\\"true\\" style=\\"position: absolute; pointer-events: none; visibility: hidden; height: 0px; overflow: hidden;\\">
                      <div class=\\"ant-tree-indent\\">
                        <div class=\\"ant-tree-indent-unit\\"></div>
                      </div>
                    </div>
                    <div style=\\"position: relative;\\" class=\\"ant-tree-list\\">
                      <div class=\\"ant-tree-list-holder\\">
                        <div>
                          <div style=\\"display: flex; flex-direction: column;\\" class=\\"ant-tree-list-holder-inner\\">
                            <div class=\\"ant-tree-treenode ant-tree-treenode-switcher-open ant-tree-treenode-leaf-last\\"><span aria-hidden=\\"true\\" class=\\"ant-tree-indent\\"></span><span class=\\"ant-tree-switcher ant-tree-switcher-noop\\"></span><span class=\\"ant-tree-checkbox\\"><span class=\\"ant-tree-checkbox-inner\\"></span></span><span title=\\"全部\\" class=\\"ant-tree-node-content-wrapper ant-tree-node-content-wrapper-normal\\"><span class=\\"ant-tree-title\\">全部</span></span></div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>"
`;
